File: Trivia_Ofer_And_Shaked_Sisso_2023/RSACryptoAlgorithm.cpp
#include "RSACryptoAlgorithm.h"
#include <random>
#include <cmath>
#include <math.h>

#define ASCCII_VLUE_LOWER_THAN_A 96

RSACryptoAlgorithm::RSACryptoAlgorithm(IDatabase* database)
    : m_database(database)
{
}

Buffer RSACryptoAlgorithm::encrypt(const Buffer& message, const int& key, const int& modulus)
{
    Buffer ecrypted;
    Byte cipherText;

    for (const Byte& ch : message)
    {
        cipherText = static_cast<Byte>(ModPow(ch-ASCCII_VLUE_LOWER_THAN_A, key, modulus));
        ecrypted.push_back(cipherText);
    }

    return ecrypted;
}

Buffer RSACryptoAlgorithm::decrypt(const Buffer& message)
{
    Buffer decrypted;
    Byte plainText;
    for (const Byte& ch : message)
    {
        if (ch == '\0')
        {
            plainText = ch;
        }
        else
        {
            plainText = static_cast<Byte>(ModPow(ch + ASCCII_VLUE_LOWER_THAN_A, m_privateKey, m_modulus));
        }
        decrypted.push_back(plainText);
    }

    return decrypted;
}

void RSACryptoAlgorithm::setDatabase(IDatabase* database)
{
    this->m_database = database;
}
void RSACryptoAlgorithm::createKeys()
{
    vector<string> keys;
    keys.reserve(2);
    int p = getRandomPrime(100, 1000);
    int q = getRandomPrime(100, 1000);
    while (p == q) {
        q = getRandomPrime(2, 100);
    }

    this->m_modulus = p * q;
    long int phi = (p - 1) * (q - 1);

    findKeys(phi, p, q);
}

void RSACryptoAlgorithm::findKeys(long int phi, long int p, long int q)
{
    int i, d;
    bool flag;
    i = getRandomPrime(2, 200);
    while (i > phi)
    {
        i = getRandomPrime(2, 200);
    }
    for (;i < phi; i++)
    {
        if (phi % i == 0)
            continue;
        flag = isPrime(i);
        if (flag && i != p && i != q)
        {
            this->m_publicKey = i;
            d = findPrivateKey(phi);
            if (d > 0)
            {
                this->m_privateKey = d;
                break;
            }
        }
    }
}

long int RSACryptoAlgorithm::findPrivateKey(long int phi)
{
    long int i = 1;
    while (true)
    {
        i += phi;
        if (i % this->m_publicKey == 0)
        {
            return (i / this->m_publicKey);
        }
    }
}

std::vector<int> RSACryptoAlgorithm::getKey()
{
    std::vector<int> keys;
    keys.push_back(this->m_publicKey);
    keys.push_back(this->m_modulus);
    return keys;
}

bool RSACryptoAlgorithm::isPrime(long int number)
{
    int i;
    double j = sqrt(number);
    for (i = 2; i <= j; i++)
    {
        if (number % i == 0)
            return false;
    }
    return true;
}

int RSACryptoAlgorithm::getRandomPrime(int min, int max)
{
    std::random_device rd;
    std::mt19937 mt(rd());
    std::uniform_int_distribution<int> dist(min, max);

    int prime = dist(mt);
    while (!isPrime(prime)) 
    {
        prime = dist(mt);
    }

    return prime;
}

long int RSACryptoAlgorithm::ModPow(int base, int exponent, int modulus)
{
    long int result = base, i = 0, j = 1;
    for (i = 0; i < exponent; i++)
    {
        j *= result;
        j %= modulus;
    }
    return result;
}

File: Trivia/RSACryptoAlgorithm.cs
using System;
using System.Collections.Generic;
using System.Numerics;
using System.Reflection;
using System.Text;
using System.Windows.Forms;

namespace Trivia
{
    class RSACryptoAlgorithm
    {
        public static int m_modulus;
        public static int m_publicKey;
        private static int m_privateKey;

        private static Random random = new Random();

        private static int GetRandomPrime(int min, int max)
        {
            Random rd = new Random();
            int prime = rd.Next(min, max + 1);
            while (!IsPrime(prime))
            {
                prime = rd.Next(min, max + 1);
            }

            return prime;
        }

        public static void CreateKeys()
        {
            List<int> keys = new List<int>(2);
            int p = GetRandomPrime(100, 1000);
            int q = GetRandomPrime(100, 1000);
            while (p == q)
            {
                q = GetRandomPrime(2, 100);
            }
            Console.WriteLine("p: " + p + " q: " + q);

            m_modulus = p * q;
            long phi = (p - 1) * (q - 1);

            FindKeys(phi, p, q);
        }
        private static void FindKeys(long phi, long p, long q)
        {
            int i, d;
            bool flag;
            i = GetRandomPrime(2, 200);
            while (i > phi)
            {
                i = GetRandomPrime(2, 200);
            }
            for (; i < phi; i++)
            {
                if (phi % i == 0)
                    continue;
                flag = IsPrime(i);
                if (flag && i != p && i != q)
                {
                    m_publicKey = i;
                    d = FindPrivateKey(phi);
                    if (d > 0)
                    {
                        m_privateKey = d;
                        break;
                    }
                }
            }
        }
        private static int FindPrivateKey(long phi)
        {
            long i = 1;
            while (true)
            {
                i += phi;
                if (i % m_publicKey == 0)
                {
                    return ((int)i / m_publicKey);
                }
            }
        }

        public static long getPublicKey()
        {
            return m_publicKey;
        }

        public static long getModulus()
        {
            return m_modulus;
        }

        public static byte[] Encrypt(byte[] plaintext, int key, int modulus)
        {
            List<byte> ecrypted = new List<byte>();

            foreach (byte ch in plaintext)
            {
                byte plainText = (byte)(ModPow((ch - 96), m_publicKey, m_modulus));
                ecrypted.Add(plainText);
            }

            return ecrypted.ToArray();
        }

        public static byte[] Decrypt(byte[] ciphertext)
        {
            List<byte> decrypted = new List<byte>();

            foreach (byte ch in ciphertext)
            {
                byte plainText = (byte)(ModPow(ch + 96, m_privateKey, m_modulus));
                decrypted.Add(plainText);
            }

            return decrypted.ToArray();
        }

        private static bool IsPrime(int number)
        {
            int i;
            double j = Math.Sqrt(number);
            for (i = 2; i <= j; i++)
            {
                if (number % i == 0)
                    return false;
            }
            return true;
        }
        private static long ModPow(int baseValue, int exponent, int modulus)
        {
            long result = baseValue;
            long j = 1;
            for (int i = 0; i < exponent; i++)
            {
                j *= result;
                j %= modulus;
            }
            return result;
        }

    }
}
